src/Bicep.Core/Syntax/SyntaxRewriteVisitor.cs

```diff
@@ -0,0 +1,534 @@
+            hasChanges |= Rewrite(syntax.Assignment, out var assignment);
+            hasChanges |= Rewrite(syntax.Value, out var value);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            return new OutputDeclarationSyntax(keyword, name, type, assignment, value);
+        }
+        void ISyntaxVisitor.VisitOutputDeclarationSyntax(OutputDeclarationSyntax syntax) => ReplaceCurrent(syntax, ReplaceOutputDeclarationSyntax);
+
+        protected virtual IdentifierSyntax ReplaceIdentifierSyntax(IdentifierSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Child, out var child);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            return new IdentifierSyntax(child);
+        }
+        void ISyntaxVisitor.VisitIdentifierSyntax(IdentifierSyntax syntax) => ReplaceCurrent(syntax, ReplaceIdentifierSyntax);
+
+        protected virtual TypeSyntax ReplaceTypeSyntax(TypeSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Identifier, out var identifier);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            return new TypeSyntax(identifier);
+        }
+        void ISyntaxVisitor.VisitTypeSyntax(TypeSyntax syntax) => ReplaceCurrent(syntax, ReplaceTypeSyntax);
+
+        protected virtual BooleanLiteralSyntax ReplaceBooleanLiteralSyntax(BooleanLiteralSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Literal, out var literal);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO calculate value correctly here
+            return new BooleanLiteralSyntax(literal, syntax.Value);
+        }
+        void ISyntaxVisitor.VisitBooleanLiteralSyntax(BooleanLiteralSyntax syntax) => ReplaceCurrent(syntax, ReplaceBooleanLiteralSyntax);
+
+        protected virtual StringSyntax ReplaceStringSyntax(StringSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.StringTokens, out var stringTokens);
+            hasChanges |= Rewrite(syntax.Expressions, out var expressions);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO calculate segmentValues correctly here
+            return new StringSyntax(stringTokens, expressions, syntax.SegmentValues);
+        }
+        void ISyntaxVisitor.VisitStringSyntax(StringSyntax syntax) => ReplaceCurrent(syntax, ReplaceStringSyntax);
+
+        protected virtual ProgramSyntax ReplaceProgramSyntax(ProgramSyntax syntax)
+        {
+            var hasChanges = Rewrite(syntax.Children, out var children);
+            hasChanges |= Rewrite(syntax.EndOfFile, out var endOfFile);
+
+            if (!hasChanges)
+            {
+                return syntax;
+            }
+
+            // TODO what to do with lexerDiagnostics?
```

Comment:
>lexerDiagnostics [](start = 36, length = 16)

This makes me wonder if we need to have errors on tokens that get collected like parse errors. (Most tokens don't have errors so it could also be a new ErrorToken class derived from Token to avoids allocating extra space for mostly null pointers.)

Reply:
I had a similar thought, but in the opposite direction - perhaps we should detach diagnostics from syntax altogether - it does become a little weird when generating syntax programmatically. There's a similar issue with spans. IIRC Rosyln doesn't attach either to syntax nodes.